README.md
#JMPX
A Stereo encoder for FM transmitters.

![Screenshot of JMPX](http://jontio.zapto.org/hda1/paradise/jmpxencoder/jmpxencoder.jpeg)

This program can be used in conjunction with any FM transmitter to produce stereo transmissions. This allows you to transmit your very own FM stereo signal which can be received with any standard FM stereo radio. The FM transmitter gets connected to the computer soundcard and JMPX takes care of the rest.

This program consists of a dynamically liked library and a GUI. The library performs the main work of obtaining audio and producing audio, and the GUI is just for control and looks. A soundcard that supports at least 96000bps is needed.

##Compiling

Qt and Qt Creator are recommended.

* Open [libJMPX/libJMPX.pro](libJMPX/libJMPX.pro) with Qt Creator and build
* Open [JMPX/JMPX.pro](JMPX/JMPX.pro) with Qt Creator and build
* Copy libJMPX.dll (or something like liblibJMPX.so.0.1 on Linux) from the build directory when building libJMPX to the build directory when building JMPX
* Run JMPX.exe (or JMPX on Linux)


Jonti 2015
http://jontio.zapto.org


kiss_fft130/CHANGELOG
//
This is same as version 1.3.0 but....

moved header stuff from kiss_fastfir.c to kiss_fastfir.h

added zero buffer testing in kiss_fastfir.c

added include block for c++ implimentation in kiss_fft.hh

removed unneeded files for real fast fir apps bar kiss_fft.hh

2015
Jonti
//

1.3.0 2012-07-18
  removed non-standard malloc.h from kiss_fft.h

  moved -lm to end of link line

  checked various return values

  converted python Numeric code to NumPy
 
  fixed test of int32_t on 64 bit OS

  added padding in a couple of places to allow SIMD alignment of structs

1.2.9 2010-05-27
  threadsafe ( including OpenMP )

  first edition of kissfft.hh the C++ template fft engine
 
1.2.8 
  Changed memory.h to string.h -- apparently more standard
  
  Added openmp extensions.  This can have fairly linear speedups for larger FFT sizes.

1.2.7 
  Shrank the real-fft memory footprint. Thanks to Galen Seitz.

1.2.6 (Nov 14, 2006) The "thanks to GenArts" release.
  Added multi-dimensional real-optimized FFT, see tools/kiss_fftndr
  Thanks go to GenArts, Inc. for sponsoring the development.

1.2.5 (June 27, 2006) The "release for no good reason" release.
   Changed some harmless code to make some compilers' warnings go away.
   Added some more digits to pi -- why not.
   Added kiss_fft_next_fast_size() function to help people decide how much to pad.
   Changed multidimensional test from 8 dimensions to only 3 to avoid testing 
   problems with fixed point (sorry Buckaroo Banzai).

1.2.4 (Oct 27, 2005)   The "oops, inverse fixed point real fft was borked" release. 
   Fixed scaling bug for inverse fixed point real fft -- also fixed test code that should've been failing.
    Thanks to Jean-Marc Valin for bug report.

   Use sys/types.h for more portable types than short,int,long => int16_t,int32_t,int64_t
   If your system does not have these, you may need to define them -- but at least it breaks in a 
   loud and easily fixable way -- unlike silently using the wrong size type.

   Hopefully tools/psdpng.c is fixed -- thanks to Steve Kellog for pointing out the weirdness.

1.2.3 (June 25, 2005)   The "you want to use WHAT as a sample" release.
    Added ability to use 32 bit fixed point samples -- requires a 64 bit intermediate result, a la 'long long'

    Added ability to do 4 FFTs in parallel by using SSE SIMD instructions. This is accomplished by
    using the __m128 (vector of 4 floats) as kiss_fft_scalar.  Define USE_SIMD to use this.
    
    I know, I know ...  this is drifting a bit from the "kiss" principle, but the speed advantages 
    make it worth it for some.  Also recent gcc makes it SOO easy to use vectors of 4 floats like a POD type.

1.2.2 (May 6, 2005)   The Matthew release
    Replaced fixed point division with multiply&shift.  Thanks to Jean-Marc Valin for 
    discussions regarding.  Considerable speedup for fixed-point.

    Corrected overflow protection in real fft routines  when using fixed point.
    Finder's Credit goes to Robert Oschler of robodance for pointing me at the bug.
    This also led to the CHECK_OVERFLOW_OP macro.

1.2.1 (April 4, 2004) 
    compiles cleanly with just about every -W warning flag under the sun

    reorganized kiss_fft_state so it could be read-only/const. This may be useful for embedded systems
    that are willing to predeclare twiddle factors, factorization.

    Fixed C_MUL,S_MUL on 16-bit platforms.

    tmpbuf will only be allocated if input & output buffers are same
    scratchbuf will only be allocated for ffts that are not multiples of 2,3,5
 
    NOTE: The tmpbuf,scratchbuf changes may require synchronization code for multi-threaded apps.


1.2 (Feb 23, 2004)
    interface change -- cfg object is forward declaration of struct instead of void*
    This maintains type saftey and lets the compiler warn/error about stupid mistakes.
            (prompted by suggestion from Erik de Castro Lopo)

    small speed improvements

    added psdpng.c -- sample utility that will create png spectrum "waterfalls" from an input file
        ( not terribly useful yet)

1.1.1 (Feb 1, 2004 )
    minor bug fix -- only affects odd rank, in-place, multi-dimensional FFTs

1.1 : (Jan 30,2004)
    split sample_code/ into test/ and tools/

    Removed 2-D fft and added N-D fft (arbitrary)

    modified fftutil.c to allow multi-d FFTs

    Modified core fft routine to allow an input stride via kiss_fft_stride()
    (eased support of multi-D ffts)

    Added fast convolution filtering (FIR filtering using overlap-scrap method, with tail scrap)

    Add kfc.[ch]: the KISS FFT Cache. It takes care of allocs for you ( suggested by Oscar Lesta ).

1.0.1 (Dec 15, 2003)
    fixed bug that occurred when nfft==1. Thanks to Steven Johnson.
    
1.0 : (Dec 14, 2003)
    changed kiss_fft function from using a single buffer, to two buffers.
    If the same buffer pointer is supplied for both in and out, kiss will
    manage the buffer copies.

    added kiss_fft2d and kiss_fftr as separate source files (declarations in kiss_fft.h )

0.4 :(Nov 4,2003) optimized for radix 2,3,4,5

0.3 :(Oct 28, 2003) woops, version 2 didn't actually factor out any radices other than 2.
        Thanks to Steven Johnson for finding this one.

0.2 :(Oct 27, 2003) added mixed radix, only radix 2,4 optimized versions

0.1 :(May 19 2003)  initial release, radix 2 only


libJMPX/JDSP.h
//---------------------------------------------------------------------------

#ifndef JDSP_H
#define JDSP_H

#include "../kiss_fft130/tools/kiss_fastfir.h"
#include "../kiss_fft130/tools/kiss_fftr.h"

#include <memory>
#include <math.h>
#include <vector>
#include <string>
#include <valarray>

#include <QDebug>

#if defined(__linux__)
#include <sys/time.h>
#include <sys/resource.h>
#include "semaphore.h"
#include <pthread.h>
#define MSSLEEP(x)	usleep(x*1000)
#else
#include <windows.h>
#define MSSLEEP(x)	Sleep(x)
#endif

#include <iostream>

#include <sstream>

#include "Definitions.h"


#define WTSIZE 8000
#define WTSIZE_1 7999

using namespace std;

class TDspGen
{
private:
        TSetGen *pSetGen;

public:

        void  ResetSettings();
         TDspGen(TSetGen *_pSetGen);
         ~TDspGen();

        vector<double> SinWT;

        double Freq;
        int SampleRate;
};



//------------------

class WaveTable
{
public:
         WaveTable(TDspGen *_pDspGen);

        ~WaveTable();
        void WTnextFrame();
        double   WTSinValue();
        double   WTSin2Value();
        void  RefreshSettings();
private:
        double WTstep;
        int intWTptr;
        double WTptr;
        TDspGen *pDspGen;
};

//-------------------

class PreEmphasis
{
private:
	double a[2];
	double b[2];
    double y[2];
	double x[2];
public:
	enum TimeConstant {WORLD,USA,NONE};
	PreEmphasis();
	void SetTc(TimeConstant timeconst);
	double Update(double val);
};

class Clipper
{
private:
	vector<double> LookupTable;
	double compressionpoint;
public:
	Clipper();
	double Update(double val);
	void SetCompressionPoint(double point);

};


class Cof_16500Hz_192000bps_119taps
{
public:
    std::vector<double> Cof;
    Cof_16500Hz_192000bps_119taps()
    {
        Cof.resize(119);
        Cof[0] = -0.000167638;Cof[1] = -0.000268748;Cof[2] = -9.56764e-05;Cof[3] = 0.000115229;Cof[4] = 0.000281049;Cof[5] = 0.000338302;Cof[6] = 0.000268326;Cof[7] = 0.000103853;Cof[8] = -8.59957e-05;Cof[9] = -0.00022598;Cof[10] = -0.000267365;Cof[11] = -0.000208623;Cof[12] = -9.46927e-05;Cof[13] = 5.66977e-06;Cof[14] = 3.5084e-05;Cof[15] = -2.04599e-05;Cof[16] = -0.00011854;Cof[17] = -0.000174899;Cof[18] = -0.00010276;Cof[19] = 0.000137925;Cof[20] = 0.000503055;Cof[21] = 0.000859024;Cof[22] = 0.00101918;Cof[23] = 0.000815281;Cof[24] = 0.000180721;Cof[25] = -0.000787676;Cof[26] = -0.0018204;Cof[27] = -0.00253487;Cof[28] = -0.00255603;Cof[29] = -0.00166446;Cof[30] = 7.89081e-05;Cof[31] = 0.00228406;Cof[32] = 0.00429358;Cof[33] = 0.00535571;Cof[34] = 0.00486947;Cof[35] = 0.00262937;Cof[36] = -0.00101614;Cof[37] = -0.00516394;Cof[38] = -0.00854399;Cof[39] = -0.00986765;Cof[40] = -0.00825062;Cof[41] = -0.00358529;Cof[42] = 0.00326712;Cof[43] = 0.010558;Cof[44] = 0.0160283;Cof[45] = 0.0175149;Cof[46] = 0.0136407;Cof[47] = 0.00439635;Cof[48] = -0.00857179;Cof[49] = -0.022094;Cof[50] = -0.0320499;Cof[51] = -0.0342823;Cof[52] = -0.0256475;Cof[53] = -0.00494015;Cof[54] = 0.026558;Cof[55] = 0.0650793;Cof[56] = 0.104962;Cof[57] = 0.139727;Cof[58] = 0.163403;Cof[59] = 0.171798;Cof[60] = 0.163403;Cof[61] = 0.139727;Cof[62] = 0.104962;Cof[63] = 0.0650793;Cof[64] = 0.026558;Cof[65] = -0.00494015;Cof[66] = -0.0256475;Cof[67] = -0.0342823;Cof[68] = -0.0320499;Cof[69] = -0.022094;Cof[70] = -0.00857179;Cof[71] = 0.00439635;Cof[72] = 0.0136407;Cof[73] = 0.0175149;Cof[74] = 0.0160283;Cof[75] = 0.010558;Cof[76] = 0.00326712;Cof[77] = -0.00358529;Cof[78] = -0.00825062;Cof[79] = -0.00986765;Cof[80] = -0.00854399;Cof[81] = -0.00516394;Cof[82] = -0.00101614;Cof[83] = 0.00262937;Cof[84] = 0.00486947;Cof[85] = 0.00535571;Cof[86] = 0.00429358;Cof[87] = 0.00228406;Cof[88] = 7.89081e-05;Cof[89] = -0.00166446;Cof[90] = -0.00255603;Cof[91] = -0.00253487;Cof[92] = -0.0018204;Cof[93] = -0.000787676;Cof[94] = 0.000180721;Cof[95] = 0.000815281;Cof[96] = 0.00101918;Cof[97] = 0.000859024;Cof[98] = 0.000503055;Cof[99] = 0.000137925;Cof[100] = -0.00010276;Cof[101] = -0.000174899;Cof[102] = -0.00011854;Cof[103] = -2.04599e-05;Cof[104] = 3.5084e-05;Cof[105] = 5.66977e-06;Cof[106] = -9.46927e-05;Cof[107] = -0.000208623;Cof[108] = -0.000267365;Cof[109] = -0.00022598;Cof[110] = -8.59957e-05;Cof[111] = 0.000103853;Cof[112] = 0.000268326;Cof[113] = 0.000338302;Cof[114] = 0.000281049;Cof[115] = 0.000115229;Cof[116] = -9.56764e-05;Cof[117] = -0.000268748;Cof[118] = -0.000167638;
    }
};


//slow fir

class FIRFilter
{
private:
    vector<double> Buf;
	int BufPtr;
public:
    vector<double> Cof;
	FIRFilter();
	double Update(double val);
    void Update(double *data,int Size);
    void UpdateInterleavedEven(double *data,int Size);
    void UpdateInterleavedOdd(double *data,int Size);
    void setBuf(double val)
    {
        Buf.resize(Cof.size());
        for(unsigned int i=0;i<Buf.size();i++)Buf[i]=val;
    }
};

//fast fir

class FastFIRFilter
{
public:
    FastFIRFilter(std::vector<kffsamp_t> &imp_responce,size_t &nfft);
    int Update(kffsamp_t *data,int Size);
    void reset();
    ~FastFIRFilter();
    size_t nfft;
    kiss_fastfir_cfg cfg;
    size_t idx_inbuf;

    std::vector<kffsamp_t> inbuf;
    std::vector<kffsamp_t> outbuf;
    std::vector<kffsamp_t> remainder;
    int remainder_ptr;

};

//stereo fast fir

class FastFIRFilterInterleavedStereo
{
public:
    FastFIRFilterInterleavedStereo(std::vector<kffsamp_t> &imp_responce,size_t &nfft);
    ~FastFIRFilterInterleavedStereo();
    void Update(kffsamp_t *data,int Size);
    void reset()
    {
        left->reset();
        right->reset();
    }
private:
    FastFIRFilter *left;
    FastFIRFilter *right;
    std::vector<kffsamp_t> leftbuf;
    std::vector<kffsamp_t> rightbuf;

};

//stereo fast fir for 16.5kHz at 192000 bps
class InterleavedStereo16500Hz192000bpsFilter
{
public:
    InterleavedStereo16500Hz192000bpsFilter()
    {
        Cof_16500Hz_192000bps_119taps lpfc;
        size_t nfft=512;
        fir = new FastFIRFilterInterleavedStereo(lpfc.Cof,nfft);
    }
    ~InterleavedStereo16500Hz192000bpsFilter()
    {
        delete fir;
    }
    void Update(kffsamp_t *data,int Size){fir->Update(data,Size);}
    void reset(){fir->reset();}
private:
    FastFIRFilterInterleavedStereo *fir;
};



#endif  //JDSP_H
